Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -47,97 +48,226 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-    if (m_requestMediaDataWhenReadySet)
-        return;
-    m_requestMediaDataWhenReadySet = true;
-    // We set requestMediaDataWhenReady only after calling enqueueSample once, to avoid having the callback
-    // being called immediately.
-    m_decompressionSession->requestMediaDataWhenReady([weakThis = ThreadSafeWeakPtr { *this }, this] {
-        if (RefPtr protectedThis = weakThis.get()) {
-            if (!--m_decodePending && m_minimumUpcomingPresentationTime) {
-                expectMinimumUpcomingSampleBufferPresentationTime(*m_minimumUpcomingPresentationTime);
-                m_minimumUpcomingPresentationTime.reset();
-            }
-            if (!m_wasNotDisplaying)
-                return;
-            m_wasNotDisplaying = false;
-            if (m_readyForMoreSampleFunction)
-                m_readyForMoreSampleFunction();
+
+    bool displaying = !sample.isNonDisplaying();
+    auto decodePromise = m_decompressionSession->decodeSample(cmSampleBuffer, displaying);
+    decodePromise->whenSettled(m_workQueue, [weakThis = ThreadSafeWeakPtr { *this }] (auto&& result) {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis)
+            return;
+
+        if (!result) {
+            ensureOnMainThread([protectedThis = WTFMove(protectedThis), status = result.error()] {
+                assertIsMainThread();
+
+                // Simulate AVSBDL decoding error.
+                RetainPtr error = [NSError errorWithDomain:@"com.apple.WebKit" code:status userInfo:nil];
+                NSDictionary *userInfoDict = @{ (__bridge NSString *)AVSampleBufferDisplayLayerFailedToDecodeNotificationErrorKey: (__bridge NSError *)error.get() };
+                [NSNotificationCenter.defaultCenter postNotificationName:AVSampleBufferDisplayLayerFailedToDecodeNotification object:protectedThis->m_renderer.get() userInfo:userInfoDict];
+                [NSNotificationCenter.defaultCenter postNotificationName:AVSampleBufferVideoRendererDidFailToDecodeNotification object:protectedThis->m_renderer.get() userInfo:userInfoDict];
+            });
+            return;
         }
+
+        if (!*result)
+            return;
+
+        RetainPtr<CMSampleBufferRef> retainedResult = *result;
+        protectedThis->decodedFrameAvailable(WTFMove(retainedResult));
     });
 }
 
 void VideoMediaSampleRenderer::initializeDecompressionSession()
 {
-    if (m_decompressionSession)
-        m_decompressionSession->invalidate();
-    m_decompressionSession = WebCoreDecompressionSession::createOpenGL();
-    m_decompressionSession->setTimebase([m_renderer timebase]);
-    m_decompressionSession->setResourceOwner(m_resourceOwner);
-    m_decompressionSession->decodedFrameWhenAvailable([weakThis = ThreadSafeWeakPtr { *this }](RetainPtr<CMSampleBufferRef>&& sample) {
-        if (RefPtr protectedThis = weakThis.get())
-            [protectedThis->m_renderer enqueueSampleBuffer:sample.get()];
+    m_workQueue->dispatchSync([&] {
```

Comment:
I think it should be up to the decompressionSession to handle its threading operation, and continue having the VideoMediaSampleRenderer being a main thread object for now.

WebCoreDecompressionSession is already design to use its own WorkQueue and all operations in parallel. There's no advantage in moving the WebCoreDecompressionSession in yet another WorkQueue.

Reply:
The advantage is not having frames hit the main thread after decoding.

Reply:
An alternative would be to keep the m_decompressionSession as-is and not worry on which thread it leaves on (it's thread-safe) and only have the m_renderer on its own thread instead . So that any `m_decompressionSession->methodWithPromise()->whenSettled(m_workQueue, [] { do_work_with_avsbdl }`

This removes the need for the sync dispatch